### PR TITLE
drivers: rtc: stm32: fix build for some L1 MCUs and allow RTC driver to work with BBRAM

### PR DESCRIPTION
**Pull request by faloj**
STM32 BBRAM depends on RTC to work. This changes STM32 RTC init stage to
PRE_KERNEL_1 to allow RTC driver to initialize before BBRAM driver.

Some adjustments are made so that kernel API is not used during the init
procedure.

Also add a test case and fix build for STM32L1 Cat.1 microcontrollers.

### CODE REVIEW

**Review by erwango — COMMENTED** [review]
Let's discuss this a bit before going on.

Counter driver is using PRE_KERNEL_1, so until now, this implementation was fine.

@semihalf-duda-patryk can you check if this is fine for you ?
Did you initially intend to get bbram up before kernel for some specific constraints ?
An alternative maybe to set RTC using PRE_KERNEL_1 as well.

**Comment by faloj** [comment]
I totaly agree that this need some discution.
Thing is if we choose to lower the init level of the RTC driver to the PRE_KERNEL_1 stage, I think some rewrite will be needed as the actual implementation depends on k_msleep (I wasn't thinking about this issue when I used it). To my understanding k_msleep do not work if the kernel has not started but I may be wrong.
Concerning the counter driver, for me raising BBRAM to POST_KERNEL won't break anything because the counter driver still start sooner.

**Comment by semihalf-duda-patryk** [comment]
> Did you initially intend to get bbram up before kernel for some specific constraints ?

Yeah. Changing PRE_KERNEL_1 to POST_KERNEL doesn't work for me.

**Comment by faloj** [comment]
I made the needed adjustemt in the RTC driver code so it can initialize a PRE_KERNEL_1 stage.

**Comment by faloj** [comment]
CI should be fixed when #62680 is merged

**Comment by faloj** [comment]
#62680 is now merged. Can someone force a CI rerun ? I don't think I can.

**Comment by erwango** [comment]
> #62680 is now merged. Can someone force a CI rerun ? I don't think I can.

Done

**Review by erwango — DISMISSED** [review]
Looks ok to me, but while you're at, would you mind adding a bbram stm32 test config using RTC to init bbram ?

**Comment by faloj** [comment]
> Looks ok to me, but while you're at, would you mind adding a bbram stm32 test config using RTC to init bbram ?

This is now done ;)

File: drivers/rtc/rtc_ll_stm32.c [hunk 1]
```diff
@@ -68,14 +70,22 @@ struct rtc_stm32_data {
 	struct k_mutex lock;
 };
 
-static int rtc_stm32_enter_initialization_mode(void)
+static int rtc_stm32_enter_initialization_mode(bool no_kernel)
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:77]: `no_kernel` as a bool is confusing. Please use `kernel_available` or similar that would avoid confusion

File: drivers/rtc/rtc_ll_stm32.c [hunk 2]
```diff
@@ -29,6 +29,10 @@
 
 LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
 
+/* Whether or not the hardware support subsecond counting */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:32]: Not related with the PR description. Please update for clarity

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/rtc/rtc_ll_stm32.c (hunk 2 above)
Comment (faloj) [reply drivers/rtc/rtc_ll_stm32.c]: This was to fix CI build which was failing while trying to build for the stm32l1_disco.

Comment (faloj) [reply drivers/rtc/rtc_ll_stm32.c]: I just realized you where probably talking about the PR description itself. This is now done.

**Review by erwango — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Pull request merged**